### PR TITLE
Fgts 162 create laser wall enemy

### PR DESCRIPTION
**Pull request by allisonchristensen**
Laser beam wall as enemy for NASA level. Beam alternates between on and off state. Player takes damage when they walk into beam.

### CODE REVIEW

**Review by fheysel — COMMENTED** [review]
This looks good to me, the one worry I have is that storing the player as a variable might be a little hack @charasyn  can you think of another work around?

File: godotproject/LaserBeam.gd [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	$Timer.start()
+	
+func _process(_delta):
+	if inside == true and AnimationState == ANIMATIONS.ATTACKING:
+		player.takeDamage(attackDamage)
+		
+# If we're in an attack state do damage
+func _on_Area2D_body_entered(body):
+	inside = true
+	player = body
```
Comment (fheysel) [inline godotproject/LaserBeam.gd]: may cause efficiency issues? I dont have any reason to back this up, just a gut feeling

Comment (fheysel) [reply godotproject/LaserBeam.gd]: Is there a way to save a pointer to the player?

Comment (charasyn) [reply godotproject/LaserBeam.gd]: I don't think this should cause issues, it seems like the best way to save a reference to the player

Comment (charasyn) [inline godotproject/LaserBeam.gd]:
```suggestion
func _on_Area2D_body_entered(body):
	if Global.is_player(body):
		inside = true
		player = body
```

File: godotproject/LaserBeam.gd [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		
+func _on_Area2D_body_exited(body):
+	inside = false
```
Comment (charasyn) [inline godotproject/LaserBeam.gd]:
```suggestion
func _on_Area2D_body_exited(body):
	if Global.is_player(body):
		inside = false
```

File: godotproject/LaserBeam.gd [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+var AnimationState = ANIMATIONS.IDLE
+var inside = false
+var player
+var attackDamage = 1
+
+# Called when the node enters the scene tree for the first time.
+func _ready():
+	$AnimatedSprite.play(AnimationStates[ANIMATIONS.IDLE])
+	print("Timer started")
+	$Timer.start()
```
Comment (charasyn) [inline godotproject/LaserBeam.gd]: Can we just set the timer off of oneshot and set it to autostart? Then we can delete $Timer.start() out of this file

File: godotproject/LaserBeam.gd [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	player = body
+		
+func _on_Area2D_body_exited(body):
+	inside = false
+
+#flips between 1 of 2 animations States
+func _on_Timer_timeout():
+	print("Timer done")
+	AnimationState = (AnimationState + 1) % ANIMATIONS.MAX_ANIMATIONS
+	$AnimatedSprite.play(AnimationStates[AnimationState])
```
Comment (charasyn) [inline godotproject/LaserBeam.gd:36]: 👍

File: godotproject/LaserBeam.tscn [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+atlas = ExtResource( 1 )
+region = Rect2( 0, 9300, 400, 300 )
+
+[sub_resource type="AtlasTexture" id=33]
+atlas = ExtResource( 1 )
+region = Rect2( 0, 9600, 400, 300 )
+
+[sub_resource type="AtlasTexture" id=34]
+atlas = ExtResource( 1 )
+region = Rect2( 0, 9900, 400, 300 )
```
Comment (charasyn) [inline godotproject/LaserBeam.tscn]: This texture is too big. Seems like it's 10200*300 which is 3.06M pixels, which is bigger than the max texture size

**Review by charasyn — CHANGES_REQUESTED** [review]
A few code changes, but more importantly, the texture for the laser is too big and needs to be shrunk

**Review by charasyn — APPROVED** [review]
Thanks for making those changes! Looks good to me 👍

**Pull request merged**